File: src/Microsoft.DotNet.Interactive/ProxyKernel.cs
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Threading.Tasks;
using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Server;

namespace Microsoft.DotNet.Interactive
{
    public sealed class ProxyKernel : 
        Kernel, 
        IKernelCommandHandler<SubmitCode>, 
        IKernelCommandHandler<RequestCompletions>,
        IKernelCommandHandler<RequestHoverText>
    {
        private readonly KernelClient _client;
        private readonly string _remoteTargetKernelName;

        public ProxyKernel(string name, KernelClient client, string remoteTargetKernelName = null) : base(name)
        {
            _client = client ?? throw new ArgumentNullException(nameof(client));
            _remoteTargetKernelName = remoteTargetKernelName;
            RegisterForDisposal(client.KernelEvents.Subscribe(OnKernelEvents));
        }

        private void OnKernelEvents(KernelEvent kernelEvent)
        {
            PublishEvent(kernelEvent);
        }

        public  Task HandleAsync(SubmitCode command, KernelInvocationContext context)
        {
            return SendCommandToRemoteKernel(command);
        }

        public Task HandleAsync(RequestCompletions command, KernelInvocationContext context)
        {
            return SendCommandToRemoteKernel(command);
        }

        public  Task HandleAsync(RequestHoverText command, KernelInvocationContext context)
        {
            return SendCommandToRemoteKernel(command);
        }

        private async Task SendCommandToRemoteKernel(KernelCommand command)
        {
            var targetKernelName = command.TargetKernelName;
            command.TargetKernelName = _remoteTargetKernelName;
            await _client.SendAsync(command);
            command.TargetKernelName = targetKernelName;
        }
    }
}

File: src/dotnet-interactive-vscode/common/interactive/proxyKernel.ts
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import * as contracts from "../interfaces/contracts";
import { Logger } from "../logger";
import { PromiseCompletionSource } from "./genericTransport";
import { IKernelCommandHandler, IKernelCommandInvocation, Kernel } from "./kernel";

export class ProxyKernel extends Kernel {

    constructor(readonly name: string, private readonly transport: contracts.Transport) {
        super(name);
    }
    getCommandHandler(commandType: contracts.KernelCommandType): IKernelCommandHandler | undefined {
        return {
            commandType,
            handle: (invocation) => {
                return this._commandHandler(invocation);
            }
        };
    }

    private async _commandHandler(commandInvocation: IKernelCommandInvocation): Promise<void> {
        const token = commandInvocation.commandEnvelope.token;
        const completionSource = new PromiseCompletionSource<boolean>();
        let sub = this.transport.subscribeToKernelEvents((envelope: contracts.KernelEventEnvelope) => {
            Logger.default.info(`proxy ${this.name} got event ${envelope.eventType} from ${envelope.command?.command?.targetKernelName} with token ${envelope.command?.token}`);
            if (envelope.command!.token === token) {
                commandInvocation.context.publish(envelope);
                if (envelope.command!.command!.id === commandInvocation.commandEnvelope.command.id) {
                    switch (envelope.eventType) {
                        case contracts.CommandFailedType:
                        case contracts.CommandSucceededType:
                            completionSource.resolve(true);
                            break;
                    }
                }
            }
        });

        try {
            this.transport.submitCommand(commandInvocation.commandEnvelope);
            Logger.default.info(`proxy ${this.name} about to await with token ${token}`);
            await completionSource.promise;
            Logger.default.info(`proxy ${this.name} done awaiting with token ${token}`);
        }
        catch (e) {
            commandInvocation.context.fail((<any>e).message);
        }
        finally {
            sub.dispose();
        }
    }
}
